File: algorithms/cracking_the_coding_interview/arrays_left_rotation.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	reader := bufio.NewReader(os.Stdin)

	lineOneString, _ := reader.ReadString('\n')
	lineOneArray := strings.Fields(lineOneString)
	lineTwoString, _ := reader.ReadString('\n')
	lineTwoArray := strings.Fields(lineTwoString)

	rotate, _ := strconv.Atoi(lineOneArray[1])
	fmt.Println(strings.Join(rotateLeft(lineTwoArray, rotate), " "))
}

func rotateLeft(a []string, rotate int) []string {
	length := len(a)
	if rotate > length {
		rotate = rotate % length
	}

	result := make([]string, length)
	for key := range a {
		var index int
		preIndex := key + rotate

		if preIndex < length {
			index = preIndex
		} else {
			index = preIndex - length
		}
		result[key] = a[index]
	}
	return result
}

File: algorithms/cracking_the_coding_interview/arrays_left_rotation.php
<?php
declare(strict_types=1);

/**
 * @param int[] $a
 * @param int $rotate
 * @return int[]
 */
function rotateLeft(array $a, int $rotate): array
{
    $length = count($a);
    if ($rotate > $length) {
        $rotate = $rotate % $length;
    }

    $result = [];
    for ($i = 0; $i < $length; $i++) {
        $preIndex = $i + $rotate;
        $index = ($preIndex < $length) ? $preIndex : ($preIndex - $length);
        $result[$i] = $a[$index];
    }
    return $result;
}

$handle = fopen ('php://stdin', 'r');
fscanf($handle, '%d %d', $n, $k);
$a = explode(' ', fgets($handle));
array_walk($a, 'intval');

echo implode(' ', rotateLeft($a, $k));
